### PR TITLE
Intel Aero uavcan support

### PR DESCRIPTION
**Pull request by anemostec**
As the Intel Aero had a CAN bus chip directly supported by it's hardware, it would have been a pity not to use it to manage CAN ESCs.
This support is now added.

### CODE REVIEW

File: libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp [hunk 1]
```diff
@@ -118,7 +118,7 @@ void AP_BoardConfig_CAN::setup_canbus(void)
             if (hal.can_mgr[drv_num - 1] != nullptr) {
                 initret &= hal.can_mgr[drv_num - 1]->begin(_var_info_can[i]._can_bitrate, i);
             } else {
-                printf("Failed to initialize can interface %d\n\r", i + 1);
+                hal.console->printf("Failed to initialize can interface %d\n\r", i + 1);
```
Comment (EShamaev) [inline libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp]: printf was there on purpose.  hal.console is not available on pixhawk when driver loads.
So I would advise putting ifdef to give different functionality to linux boards.

File: libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp [hunk 2]
```diff
@@ -129,7 +129,7 @@ void AP_BoardConfig_CAN::setup_canbus(void)
         for (uint8_t i = 0; i < MAX_NUMBER_OF_CAN_DRIVERS; i++) {
             if (hal.can_mgr[i] != nullptr) {
                 hal.can_mgr[i]->initialized(true);
-                printf("can_mgr %d initialized well\n\r", i + 1);
+                hal.console->printf("can_mgr %d initialized well\n\r", i + 1);
```
Comment (EShamaev) [inline libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp]: see above

File: libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp [hunk 3]
```diff
@@ -144,7 +144,7 @@ void AP_BoardConfig_CAN::setup_canbus(void)
                         if (_var_info_can_protocol[i]._uavcan->try_init() == true) {
                             any_uavcan_present = true;
                         } else {
-                            printf("Failed to initialize uavcan interface %d\n\r", i + 1);
+                            hal.console->printf("Failed to initialize uavcan interface %d\n\r", i + 1);
```
Comment (EShamaev) [inline libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp]: see above

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp [hunk 4]
```diff
@@ -26,6 +26,7 @@
 #include <uavcan/equipment/actuator/Command.hpp>
 #include <uavcan/equipment/actuator/Status.hpp>
 #include <uavcan/equipment/esc/RawCommand.hpp>
+#include <uavcan/equipment/esc/Status.hpp>
```
Comment (EShamaev) [inline libraries/AP_UAVCAN/AP_UAVCAN.cpp]: Why is this needed?

File: libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp (hunk 1 above)
Comment (anemostec) [reply libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp]: ok, maybe it's easier to revert that change?
I'll put printf back in place as printf is also supported on all linux boards

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp (hunk 4 above)
Comment (anemostec) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: I put the reason in the commit text. This is to have in the future an access to ESC current consumption, voltage, rpm, node id, analog input ...

Comment (EShamaev) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: Then please add it when it will be used.

File: libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp (hunk 1 above)
Comment (EShamaev) [reply libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp]: If that is OK on linux boards - then yes.

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp (hunk 4 above)
Comment (anemostec) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: ok, I'll cancel that change

Comment (anemostec) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: revert pushed

File: libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp (hunk 1 above)
Comment (anemostec) [reply libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp]: revert pushed

File: libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp (hunk 2 above)
Comment (anemostec) [reply libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp]: done

File: libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp (hunk 3 above)
Comment (anemostec) [reply libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp]: done

File: libraries/AP_UAVCAN/AP_UAVCAN.cpp (hunk 4 above)
Comment (anemostec) [reply libraries/AP_UAVCAN/AP_UAVCAN.cpp]: revert pushed

**Comment by EShamaev** [comment]
If it was fly tested - looks fine after squashing the commits.

**Comment by anemostec** [comment]
Hello, does that mean this PR will be merged soon? or do you expect me to make some more changes or checks. We will implement the aero PPM support after this PR merge and make our drone fly with it.

**Comment by EShamaev** [comment]
Please squash commits

**Comment by anemostec** [comment]
done

**Comment by EShamaev** [comment]
Please squash the commits and make it as outlined here:
http://ardupilot.org/dev/docs/submitting-patches-back-to-master.html

**Comment by anemostec** [comment]
I don't get it, your page does only explain how to create a new PR and submit it to the github.
I regrouped all commits into only one and squashed the others as you wanted.

**Comment by lucasdemarchi** [comment]
Did you do your own extension board? CAN pins weren't exposed when I was working with it, but I flew it with can ESCs.

**Comment by anemostec** [comment]
@lucasdemarchi yes we have our own extension board as the CAN pins were not exposed by default. How did you manage to use CAN ESCs without your own extension board?

**Comment by lucasdemarchi** [comment]
@anemostec I never said I didn't have ;)

**Comment by anemostec** [comment]
oh, so you had also your own CAN driver code? Why didn't you incorporate it to the ardupilot code?
That would have spared us some effort ;)

**Comment by lucasdemarchi** [comment]
Not enough time, sorry for that.

**Comment by khancyr** [comment]
@anemostec you need to get ride of merge and revert commit to keep the history clean ! Look at the commit history to see how it should be !
In fact, you go only 3 commits : 
AP_HAL: Intel Aero: add define HAL_WITH_UAVCAN
waf: add support UAVCAN for Intel Aero board
AP_HAL: RCOutput_Tap: Add support for CAN driver (or something like that)

But aren't can ESC already support by linux plateform ? So should the uavcan esc output be written in only one place ?

**Comment by anemostec** [comment]
@khancyr I'm afraid I'm not very used to git. I will ask the forum for help for this.
Concerning CAN ESC, it was not supported for the Intel Aero board. Only the generic CAN access was available for all linux boards. And yes this esc output should be written maybe somewhere else. Maybe @EShamaev could point out a more adequate location, I'll make the change according to his advice.

**Comment by EShamaev** [comment]
OK, I will take a deeper look.
From what I have seen for Aero the communication with ESC was relocated to other place globally.
Maybe we can find the one who did this support initially?

**Comment by anemostec** [comment]
I think @lucasdemarchi did the aero support initially

**Comment by anemostec** [comment]
@khancyr is the commit history better now? I tried to follow the tutorial I found on git rebase -i

**Comment by khancyr** [comment]
yep good! But you lose the waf commit ... isn't it needed ?

**Comment by anemostec** [comment]
don't we still have the ardupilotwaf commit? I still have it in my local history

**Comment by khancyr** [comment]
ho, it is one top ! my bad

**Comment by lucasdemarchi** [comment]
@magicrub as I said the tap driver should not be changed to include uavcan . This is in the wrong place as is the current code in PX4RCOutput. The problem here is that it will most likely fail due to the uart being used somewhere else.

**Comment by EShamaev** [comment]
I will look into both places this weekend.

**Comment by anemostec** [comment]
@EShamaev hello, any chance to look at it this WE?

**Comment by EShamaev** [comment]
Closed in favor of https://github.com/ArduPilot/ardupilot/pull/7882
